### DISCUSSION TITLE
Logseq export is coming

### DISCUSSION THREAD

**Discussion by ahyatt** [ahyatt/ekg#22] [Announcements]
Hi all, 

If you want to be the first on your block to try out ekg's export to logseq, try it out in the `logseq` branch.  

```
(require 'ekg-logseq)
;; Set the parent logseq directory
(setq ekg-logseq-dir "~/logseq")
```

Then, you can run `M-x ekg-logseq-export`.

There's more I need to test out, and I'm not sure I'm happy with the choices I had to make, but it does actually work.

**Comment by ahyatt** [comment]
This is now merged into the `develop` branch.  It should more or less work, although I'd make backups before you try.  Note especially that this overwrites pages in logseq in which you may have existing data not in ekg.  This is really meant for people who want to just mirror data to logseq to read what they have in ekg.

**Comment by jayrajput** [comment]
I did tried the `ekg-logseq-export` and works fine for my use case. I want to test it more though. One thing I want to do is to have this operation being blazing fast. Can we add some kind of output which shows the time taken by this command. I am hoping that the subsequent execution shall be faster.

The use case I have is to export the updated `ekg-notes` at the end of the day and then using git to visually review the changes and push it to my personal git repo.

**Reply by ahyatt** [reply]
Thanks for trying!  I'm not sure it can get much faster, but I can look into it.  Adding timing seems reasonable, I'll add that soon.  What I can also do is make it execute in its own thread, so it won't lock up your emacs. 

What I'd recommend to do, without waiting for me, is calling `profiler-start`, running the command, then `profiler-stop`.  Then `profiler-report` should give you lots of interesting timing information, including the time spent on the whole thing.

**Reply by jayrajput** [reply]
Thanks. I will give profiler a try.

**Reply by jayrajput** [reply]
Tried the profiler and read the code. I realized that the export is not incremental. Basically it process all the tags each time. Would not it be nice to process only the tags/notes which were updated since the last export?

**Reply by jayrajput** [reply]
Btw, things are not terribly slow. It is less than a minutes for some 1000+ org-roam which I have exported to ekg-notes and now to logseq. But since this export is what I plan to do every day to make sure I have a git-based file backup, I want it to be as fast as possible.

**Reply by ahyatt** [reply]
I've added functionality to import and export from the last time, as you suggested. This required a change to the `triples` library, so please update to the newest version, which is available on GNU ELPA.

**Reply by jayrajput** [reply]
Things are blazing fast in my testing.

Observation without changes

```
; Execution without changes.
; first execution
(benchmark-call #'ekg-logseq-export 1)
(33.935868 191 15.481697999999994)

; multiple more execution without any new ekg-notes
(benchmark-call #'ekg-logseq-export 3)
(113.540648 478 47.21308800000001)

; created a dummy ekg-notes and ran again
(benchmark-call #'ekg-logseq-export 1)
(35.502831 136 15.476445000000012)
```

Measurement with changes

```
; Execution with changes
; first execution 
(benchmark-call #'ekg-logseq-export 1)
(52.3588 1283 32.662945)

; second execution without any new ekg-notes
(benchmark-call #'ekg-logseq-export 1)
(0.501267 7 0.21040199999999487)

; third execution with a dummy new ekg-note
(benchmark-call #'ekg-logseq-export 1)
(0.552199 8 0.2540520000000015)
```

**Comment by jayrajput** [comment]
Since I am not an logseq user. Every time I export the ekg-notes to logseq, there is a cognitive load of figuring out the changes. A simple export where each note is export to a file would have been better for my use cases of taking file backups of my notes in a git repository.

I keep on thinking of exporting the ekg-notes to denote format files with/without the front-matter added by denote. Probably a better denote. Denote does support creating the files without the titles. Denote does require the timestamp upto  granularity of second and I wonder if the note creation time upto the granularity of second is already available in the triples database.

**Comment by ahyatt** [comment]
Yes, we do have note creation time in second granularity.  It is stored in the `time-tracked` type, in this case the property is `time-tracked/creation-time`.  We use `time-tracked/modification-time` in the logseq export to understand what to export.

Exporting to different things such as denote can be done.  It'd be interesting to understand why, though.  Logseq export, for me, gets me something I can't get with ekg, which is being able to read and write notes on my phone.

**Comment by jayrajput** [comment]
Thanks for the reply.

My use cases is simply to see what was changed (added/edited/deleted) from my notes at the end of the day and then push the notes in my git repo. For this simple use cases, denote seems to work better as the ekg-note can be one to one mapped with denote-note.

With `logseq`, a single ekg-note change may trigger multiple logseq note change.

The denote exporter may also provide an easy code sample for someone to read and write their own exporter/importer.

A side-effect of this is that it provide easy access to denote note which denote does not provide by default. There are other packages like `denote-menu` which has to be used for easy browsing of denote notes.

Btw, there is one challenge which I do see. Denote allows only limited characters in their tags whereas `ekg` has no such limitation. User  who desires to use both `ekg` and `denote` will have to manually take care of using the appropriate tag name. The challenge though come with the prefix tag which have the slash `/` hard-coded. I think `ekg` should make the separator for prefix tag as configurable for user who prefer to use `ekg` and `denote` together. I as a user do not want to miss on the prefix tag feature.

**Comment by ahyatt** [comment]
Yes, that seems like a pretty reasonable and interesting use-case to me.  I don't think I will be writing a denote exporter in the next month at least, so I'd encourage you to write one if it seems useful to you, following the patterns in the `ekg-logseq` module.

**Reply by jayrajput** [reply]
Thanks. I will give that a try. I will be looking for some guidance though. 

Although I am quietly looking more into logseq and trying to understand it more for other use case like usage on phone. Overall `logseq` also looks like an interesting use case.
